src/main/java/org/springframework/data/neo4j/core/mapping/PropertyTraverser.java

```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright 2011-2021 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.data.neo4j.core.mapping;
+
+import java.util.Comparator;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.TreeSet;
+import java.util.function.BiConsumer;
+import java.util.function.BiPredicate;
+
+import org.apiguardian.api.API;
+import org.springframework.data.mapping.Association;
+import org.springframework.data.mapping.PropertyPath;
+import org.springframework.lang.Nullable;
+
+/**
+ * A strategy for traversing all properties (including association) once, without going in circles with cyclic mappings.
+ * Uses the same idea of relationship isomorphism like Cypher does (Relationship isomorphism means that one relationship
+ * or association cannot be returned more than once for each entity).
+ *
+ * @author Michael J. Simons
+ * @since 6.3
+ */
+@API(status = API.Status.INTERNAL)
+public final class PropertyTraverser {
+
+	private final Neo4jMappingContext ctx;
+	private final Set<Association<?>> pathsTraversed = new HashSet<>();
+
+	public PropertyTraverser(Neo4jMappingContext ctx) {
+		this.ctx = ctx;
+	}
+
+	public void traverse(
+			Class<?> root,
+			BiConsumer<PropertyPath, Neo4jPersistentProperty> sink
+	) {
+		traverse(root, (path, toProperty) -> true, sink);
+	}
+
+	public synchronized void traverse(
+			Class<?> root,
+			BiPredicate<PropertyPath, Neo4jPersistentProperty> predicate,
+			BiConsumer<PropertyPath, Neo4jPersistentProperty> sink
+	) {
+		this.pathsTraversed.clear();
```

Comment:
Given this call and the current usage looks like `new PropertyTraverser(..).traverse(..)` wouldn't it make sense to make those methods _static_?

Reply:
No. The thing has state.